From 72c7a629334822a96a787c89d5a510f15a833708 Mon Sep 17 00:00:00 2001
From: Brent Cook <busterb@gmail.com>
Date: Fri, 27 Mar 2015 23:14:15 -0500
Subject: [PATCH 09/21] Notify the user when constraint support is disabled.

Update the manpage and warn if constraints are
configured but ntpd is built without libtls present.
From Paul B. Henson.
---
 src/usr.sbin/ntpd/config.c     |  3 +++
 src/usr.sbin/ntpd/constraint.c |  2 ++
 src/usr.sbin/ntpd/ntpd.conf.5  | 11 +++++++++--
 3 files changed, 14 insertions(+), 2 deletions(-)

diff --git a/src/usr.sbin/ntpd/config.c b/src/usr.sbin/ntpd/config.c
index e243818c255..856c3147cc5 100644
--- a/src/usr.sbin/ntpd/config.c
+++ b/src/usr.sbin/ntpd/config.c
@@ -187,6 +187,9 @@ new_constraint(void)
 	p->id = ++constraint_maxid;
 	p->fd = -1;
 
+#ifndef HAVE_LIBTLS
+	log_warnx("constraint configured without libtls support");
+#endif
 	return (p);
 }
 
diff --git a/src/usr.sbin/ntpd/constraint.c b/src/usr.sbin/ntpd/constraint.c
index 7729c168c53..8f54597a066 100644
--- a/src/usr.sbin/ntpd/constraint.c
+++ b/src/usr.sbin/ntpd/constraint.c
@@ -353,10 +353,12 @@ priv_constraint_child(const char *pw_dir, uid_t pw_uid, gid_t pw_gid)
 	if (setpriority(PRIO_PROCESS, 0, 0) == -1)
 		log_warn("could not set priority");
 
+#ifdef HAVE_LIBTLS
 	/* load CA certs before chroot() */
 	if ((conf->ca = tls_load_file(tls_default_ca_cert_file(),
 	    &conf->ca_len, NULL)) == NULL)
 		fatalx("failed to load constraint ca");
+#endif
 
 	if (chroot(pw_dir) == -1)
 		fatal("chroot");
diff --git a/src/usr.sbin/ntpd/ntpd.conf.5 b/src/usr.sbin/ntpd/ntpd.conf.5
index ecc52fe6964..07b731db866 100644
--- a/src/usr.sbin/ntpd/ntpd.conf.5
+++ b/src/usr.sbin/ntpd/ntpd.conf.5
@@ -216,8 +216,15 @@ authenticated constraint,
 thereby reducing the impact of unauthenticated NTP
 man-in-the-middle attacks.
 Received NTP packets with time information falling outside of a range
-near the constraint will be discarded and such NTP servers
-will be marked as invalid.
+near the constraint will be discarded and such NTP servers will be marked as
+invalid.
+.Pp
+Support for constraints is only available if
+.Xr ntpd 8
+has been linked with libtls from LibreSSL. Configuring a constraint
+without libtls causes
+.Xr ntpd 8
+to log a warning message on startup.
 .Bl -tag -width Ds
 .It Ic constraint from Ar url [ip...]
 Specify the URL, IP address or the hostname of an HTTPS server to
-- 
2.42.0

